### PR TITLE
Refactor sequence matching to accept item rather than word array

### LINKED ISSUE #8810: Quickbar doesn't show options in other language
### Checklist

I have updated to the latest available Home Assistant version.
I have cleared the cache of my browser.
I have tried a different browser to see if it is related to my browser.

### Describe the issue you are experiencing

The quickbar doesn't show script reloading in my language (German).
If i switch to English, it seems all options are shown. 
See screenshots below.

### Describe the behavior you expected

Quickbar should show all options in all languages.

### Steps to reproduce the issue

1. Set UI to german, Reload browser
2. Open quickbar, type 'script'
3. Change UI to englisch, Reload browser
4. OpOpen quickbar, type 'script'

### What version of Home Assistant Core has the issue?

core-2021.4.0b4

### What was the last working version of Home Assistant Core?

_No response_

### In which browser are you experiencing the issue with?

Latest FF and Chrome

### Which operating system are you using to run this browser?

Ubuntu 18.03

### State of relevant entities

```yaml
# Paste your state here.

```

### Problem-relevant frontend configuration

```yaml
# Paste your YAML here.

```

### Javascript errors shown in your browser console/inspector

```txt
# Paste your logs here.

```

UI setting  German:
Auswahl_457

UI setting English
Auswahl_458

### PR DESCRIPTION
**Pull request by donkawechico**
## Proposed change

Preparation for quick-bar letter highlighting (Draft PR here: https://github.com/home-assistant/frontend/pull/8868). Changes the `fuzzyFilterSort` method to accept the whole item rather than an array of filter words.

This has a few benefits:

1. Allows items to add an arbitrary number of "words" to represent it during filtration rather than just two.
2. Makes it so the sorting method doesn't need to know about "altText" or "filterText".
3. The sorter needs access to the item so that it can give the item a text decoration (for eventual letter highlighting). Without that, we'd have to loop over all the items a second time and apply the decoration algo.
4. Might help resolve https://github.com/home-assistant/frontend/issues/8810

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/common/string/filter/sequence-matching.ts [hunk 1]
```diff
@@ -51,8 +54,7 @@ export const fuzzySequentialMatch = (filter: string, ...words: string[]) => {
 
 export interface ScorableTextItem {
   score?: number;
-  filterText: string;
-  altText?: string;
+  words: string[];
```
Comment (donkawechico) [inline src/common/string/filter/sequence-matching.ts:57]: I'm open to other names for this property. All the names I could think of didn't seem quite right. 

"words" isn't quite right cuz it's not a list of single words. It's a list of target strings that our filter will be trying to match against.

"filterStrings" maybe? Or "stringIdentifiers"?

Comment (bramkragten) [reply src/common/string/filter/sequence-matching.ts:57]: Just `strings`?

File: src/common/string/filter/sequence-matching.ts [hunk 2]
```diff
@@ -10,10 +10,13 @@ import { fuzzyScore } from "./filter";
  * @return {number} Score representing how well the word matches the filter. Return of 0 means no match.
  */
 
-export const fuzzySequentialMatch = (filter: string, ...words: string[]) => {
+export const fuzzySequentialMatch = (
+  filter: string,
+  item: ScorableTextItem
```
Comment (bramkragten) [inline src/common/string/filter/sequence-matching.ts:15]: Why this changes? We dont use the item for anything else here?

File: src/common/string/filter/sequence-matching.ts [hunk 3]
```diff
@@ -63,9 +65,7 @@ type FuzzyFilterSort = <T extends ScorableTextItem>(
 export const fuzzyFilterSort: FuzzyFilterSort = (filter, items) => {
   return items
     .map((item) => {
-      item.score = item.altText
-        ? fuzzySequentialMatch(filter, item.filterText, item.altText)
-        : fuzzySequentialMatch(filter, item.filterText);
+      item.score = fuzzySequentialMatch(filter, item);
```
Comment (bramkragten) [inline src/common/string/filter/sequence-matching.ts:68]: Could just do:
```suggestion
      item.score = fuzzySequentialMatch(filter, item.words);
```

File: src/dialogs/quick-bar/ha-quick-bar.ts [hunk 4]
```diff
@@ -228,9 +233,15 @@ export class QuickBar extends LitElement {
...
 
   private _renderItem(item: QuickBarItem, index?: number) {
-    return isCommandItem(item)
-      ? this._renderCommandItem(item, index)
-      : this._renderEntityItem(item, index);
+    if (isCommandItem(item)) {
+      return this._renderCommandItem(item, index);
+    }
+
+    if (isEntityItem(item)) {
```
Comment (bramkragten) [inline src/dialogs/quick-bar/ha-quick-bar.ts:240]: This is just... we check if it is not a command item, we already know that, because we checked that above...?

File: src/dialogs/quick-bar/ha-quick-bar.ts [hunk 5]
```diff
@@ -412,32 +419,35 @@ export class QuickBar extends LitElement {
...
 
       return {
-        primaryText,
-        filterText: `${categoryText} ${primaryText}`,
-        action: () => this.hass.callService(domain, "reload"),
+        ...commandItem,
         categoryKey: "reload",
-        iconPath: mdiReload,
-        categoryText,
+        words: [`${commandItem.categoryText} ${commandItem.primaryText}`],
```
Comment (bramkragten) [inline src/dialogs/quick-bar/ha-quick-bar.ts:450]: Why not this?
```suggestion
        words: [commandItem.categoryText, commandItem.primaryText],
```

Comment (donkawechico) [reply src/dialogs/quick-bar/ha-quick-bar.ts:450]: Say a user types "relaut"  in the quickbar, hoping to find "Reload Automations".

If we send as word array, then it would check "relaut" against "Reload" and fail. Then it will check "relaut" against "Automations" and fail.

File: src/dialogs/quick-bar/ha-quick-bar.ts (hunk 4 above)
Comment (donkawechico) [reply src/dialogs/quick-bar/ha-quick-bar.ts:240]: `_renderEntityItem(item)` expects an `EntityItem`, but `item` is a `QuickBarItem` here. This would be okay except that the `EntityItem` interface now has the `altText` property.

So I'm no longer allowed to blindly send a `QuickBarItem` to the `_renderEntityItem` method due to mismatching types.

File: src/common/string/filter/sequence-matching.ts (hunk 3 above)
Comment (donkawechico) [reply src/common/string/filter/sequence-matching.ts:68]: Unfortunately not. The point here is that we need `fuzzySequentialMatch` to have access to the whole `item` so that we can efficiently calculate and store the "decorated text" (for letter highlighting) on the item itself during the fuzzy checking loop.

This is to avoid having to loop over all the matches after filtering to get the text decoration.

File: src/common/string/filter/sequence-matching.ts (hunk 1 above)
Comment (donkawechico) [reply src/common/string/filter/sequence-matching.ts:57]: I'm okay with just `strings`. It's just not very self-descriptive. I could rectify that in the commented method descriptor though.

File: src/common/string/filter/sequence-matching.ts (hunk 2 above)
Comment (donkawechico) [reply src/common/string/filter/sequence-matching.ts:15]: See other comment. When we do letter highlighting, we want to store the text decorations on each item as their match scores are being calculated during filtering.

Here's the relevant line of code in the PR for letter highlighting where we will use `item`:
https://github.com/home-assistant/frontend/pull/8868/files#diff-692eb3ca3265b7ffde8a1d7e15fc059f21c80d88ef1105fd046e0a0c6ad3da4cR82

File: src/dialogs/quick-bar/ha-quick-bar.ts (hunk 4 above)
Comment (bramkragten) [reply src/dialogs/quick-bar/ha-quick-bar.ts:240]: `as EntityItem` ?

Comment (donkawechico) [reply src/dialogs/quick-bar/ha-quick-bar.ts:240]: Could do, but I try to avoid `X as Y` as it [weakens type safety](https://medium.com/swlh/why-you-should-avoid-type-assertions-in-typescript-5494e3d04dd).

File: src/common/string/filter/sequence-matching.ts (hunk 1 above)
Comment (donkawechico) [reply src/common/string/filter/sequence-matching.ts]: Changed to "strings" and adjust the tsdoc descriptions to help explain the parameters better.
image

File: src/dialogs/quick-bar/ha-quick-bar.ts (hunk 4 above)
Comment (bramkragten) [reply src/dialogs/quick-bar/ha-quick-bar.ts:240]: yes, but I rather do that than to add unneeded runtime checks 😄

Comment (donkawechico) [reply src/dialogs/quick-bar/ha-quick-bar.ts:240]: Okie doke

**Review by bramkragten — APPROVED** [review]

File: src/dialogs/quick-bar/ha-quick-bar.ts (hunk 4 above)
Comment (donkawechico) [reply src/dialogs/quick-bar/ha-quick-bar.ts]: Fixed
